### ISSUE TITLE
BlockBlobService.get_blob_to_path() crashes even when if_match='*'

### ISSUE THREAD

**Issue by marco-rossi29** [Azure/azure-storage-python#625]
### Which service(blob, file, queue) does this issue concern?
BlockBlobService.get_blob_to_path()

### Which version of the SDK was used? Please provide the output of `pip freeze`.
Same issue is encountered in versions: 0.37.1, 1.3.0, and 2.1.0

**Version 0.37.1**
>pip freeze
azure-common==1.1.8
azure-nspkg==2.0.0
azure-storage==0.36.0
azure-storage-blob==0.37.1
azure-storage-common==0.37.1
azure-storage-nspkg==2.0.0

**Version 1.3.0**
>pip freeze
azure-common==1.1.8
azure-nspkg==2.0.0
azure-storage==0.36.0
azure-storage-blob==1.3.0
azure-storage-common==1.3.0
azure-storage-nspkg==2.0.0

**Latest version**
>pip freeze
azure-common==1.1.8
azure-nspkg==2.0.0
azure-storage==0.36.0
azure-storage-blob==2.1.0
azure-storage-common==2.1.0
azure-storage-nspkg==2.0.0

### What problem was encountered?
When trying to download a blob that is currently been written in append mode (i.e., the ETag changes during time of download), `BlockBlobService.get_blob_to_path()` crashes even when using `if_match='*'` giving **ConditionNotMet** Error:
`Client-Request-ID=9cce10c8-b739-11e9-9366-b32a86ae676b Retry policy did not allow for a retry: Server-Timestamp=Mon, 05 Aug 2019 04:29:28 GMT, Server-Request-ID=86bb1780-a01e-002e-6146-4b01e3000000, HTTP status code=412, Exception=The condition specified using HTTP conditional header(s) is not met. ErrorCode: ConditionNotMet<?xml version="1.0" encoding="utf-8"?><Error><Code>ConditionNotMet</Code><Message>The condition specified using HTTP conditional header(s) is not met.RequestId:86bb1780-a01e-002e-6146-4b01e3000000Time:2019-08-05T04:29:28.3320963Z</Message></Error>`

### Have you found a mitigation/solution?
- Using a version before **1.3.1** and using `max_connections=1` mitigates the problem
- Downloading a very small portion of the blob (< 20 MB) will work since ETag doesn't change during time of download

Code to repro:
```
bbs = BlockBlobService(connection_string=connection_string)
bbs.get_blob_to_path(app_id, blob_name, output_fp, start_range=0, end_range=100*1024**2, if_match='*')
```

**Comment by zezha-msft** [comment]
Hi @marco-rossi29, thanks for reaching out!

This behavior is by design, as the SDK needs to guarantee the data integrity: once the transfer starts, the downloader will lock onto the blob's Etag, as we cannot possibly know which part of the blob changed. 

When given if_match='*', the method checks the condition by performing a GET request on the beginning of the blob; and the Etag to lock onto is obtained in the same call.

You could work around this problem by having your own version of the downloader, would that be a satisfactory solution for you?

**Comment by marco-rossi29** [comment]
yes, I'm happy to have my own version of the downloader. What should I change in the code?

**Comment by zezha-msft** [comment]
option 1 seems to be the easiest. 😄

**Comment by marco-rossi29** [comment]
Thank you for your answer. I'm now puzzeled by the `if_match='*'` specified in the docs:

> if_match (str) – An ETag value, or the wildcard character (*). Specify this header to perform the operation only if the resource’s ETag matches the value specified.

My expectation is that when using `if_match='*'` the blob is downloaded no matter what the ETag is. But, as we saw above, this is not the case.

Can you give me an example of when using `if_match='*'` would give a behaviour different than the default, i.e., `if_match=None`?

**Comment by zezha-msft** [comment]
`if_match='*'` guarantees that a blob exists (i.e. there is an actual Etag to look at). `if_none_match='*'` guarantees that a blob does not exist.

**Comment by marco-rossi29** [comment]
I see. Thanks. So the bottom line is that the current API doesn't allow me to download a file while it is modified by an append operation. Assuming the user takes responsibility about the integrity of the file, is it possible to add a flag or any other mechanism on your side to ensure that get_blob_to_*() would work in this append scenario?

**Comment by marco-rossi29** [comment]
Would something like this be acceptable:
[](https://github.com/Azure/azure-storage-python/pull/630)
